### DISCUSSION TITLE
How to do action after create Filament Resource

### DISCUSSION THREAD

**Discussion by MahmoudAdelMansour** [filamentphp/filament#11999] [Help]
### Package

Form builder

### Package Version

v3

### How can we help you?

i use filament resource for Product Resource , and i want to send newsletter or do anything after record , 
i see there's ->after or ->before or ->action method's but i try do use it with form but i don't how to do it , 
i know there's a solution to do that with model , but my task is to do that at filament resource with create action ,
any idea ?

**Comment by dissto** [comment]
I suppose you could use a [lifecycle hook](https://filamentphp.com/docs/3.x/panels/resources/creating-records#lifecycle-hooks) in your `CreateProduct` class.

```php
protected function afterCreate(): void
{
  // do what you need to do to send the newsletter. Perhaps dispatch a job
  SendNewsletter::dispatch();
}
```

Or if you prefer to use a [lifecycle hook](https://filamentphp.com/docs/3.x/actions/prebuilt-actions/create#lifecycle-hooks) on the `CreateAction` itself

```php
CreateAction::make()
->after(function () {
    SendNewsletter::dispatch();
})
```

**Reply by MahmoudAdelMansour** [reply]
Thanks for the reply. I tried to use it, but it didn't have any effect. I wrote it inside the Resource, is that correct?

**Reply by leandrocfe** [reply]
`ProductResource/Pages/CreateProduct.php` actually.
https://filamentphp.com/docs/3.x/panels/resources/creating-records#lifecycle-hooks

`ProductResource/Pages/ManageProducts.php` if you are using a simple resource. 
https://filamentphp.com/docs/3.x/actions/prebuilt-actions/create#lifecycle-hooks

**Reply by abdulmannans-joistic** [reply]
How to overrirde the default New Resource button and use lifecycle hooks with it? @leandrocfe

**Reply by leandrocfe** [reply]
Are you using a simple resource?
https://filamentphp.com/docs/3.x/actions/prebuilt-actions/create#lifecycle-hooks

**Reply by abdulmannans-joistic** [reply]
Yes I am using simple resource. And have modal button on default on the top. How can I make few changes or perform any action before or after saving data.

**Reply by abdulmannans-joistic** [reply]
<img width="1118" alt="Screenshot 2024-04-04 at 5 26 21 PM" src="https://github.com/filamentphp/filament/assets/162964040/07ac7f4a-38f5-484d-b111-d53a20b3ce76">

I want to customize this default action. Is there any way?

**Reply by leandrocfe** [reply]
Yes, this action corresponds to a `CreateAction` available in the ListPage or ManagePage of your Resource.

**Reply by abdulmannans-joistic** [reply]
Thank you @leandrocfe I have found it.

**Reply by rahadianap** [reply]
did you find it? @abdulmannans-joistic.

**Reply by abdulmannans** [reply]
@rahadianap yes i used using which let me customize it.

**Reply by fahmiegerton** [reply]
> Thank you @leandrocfe I have found it.

I couldn't find it, mind to share where to look at or any tutorial how to do it? Thanks!

**Reply by abdulmannans-joistic** [reply]
@fahmiegerton  this is how I have implemented 

```
use Illuminate\Database\Eloquent\Model;
 
CreateAction::make()
    ->using(function (array $data, string $model): Model {
        $record = $model::create($data);
        // action that needs to be done 
        return $record;
    })
```

**Reply by fahmiegerton** [reply]
> @fahmiegerton this is how I have implemented
> 
> ```
> use Illuminate\Database\Eloquent\Model;
>  
> CreateAction::make()
>     ->using(function (array $data, string $model): Model {
>         $record = $model::create($data);
>         // action that needs to be done 
>         return $record;
>     })
> ```

May I know where do you put this?

**Reply by abdulmannans-joistic** [reply]
@fahmiegerton 
You can use it in header actions and in table actions as well. But do mind this is only for simple resource. If you are using pages use lifecycle hooks.
